### 全连接网络分类 MNIST 数据集

**数据准备**
+ 读取MNIST训练和测试数据
+ 图像像素范围有$[0,255]$变换到$[0,1]$
+ 显示训练集和测试集大小

In [1]:
import numpy as np
import pandas as pd

train_data = pd.read_csv("MNIST_train.csv")
train_images = train_data.iloc[:,1:785].to_numpy()
train_images = train_images.astype('float32')/255.

test_data = pd.read_csv("MNIST_test.csv")
test_images = test_data.iloc[:,1:785].to_numpy()
test_images = test_images.astype('float32')/255.

print("Shape of train samples:", train_images.shape)
print("Shape of test samples:", test_images.shape)

Shape of train samples: (60000, 784)
Shape of test samples: (10000, 784)


**数据准备**
+ 将训练数据和测试数据的label转换为one-hot编码

In [20]:
from tensorflow.keras.utils import to_categorical

train_labels = train_data.iloc[:,0].to_numpy()
test_labels = test_data.iloc[:,0].to_numpy()
print("Original labels:\n", train_labels[0:10])

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

print("One-hot labels:\n", train_labels[0:10,:])
print("\nShape of train labels:", train_labels.shape)
print("Shape of test labels:", test_labels.shape)

Original labels:
 [5 0 4 1 9 2 1 3 1 4]
One-hot labels:
 [[0. 0. 0. 0. 0. 1. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]
 [0. 0. 1. 0. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 1. 0. 0. 0. 0. 0.]]

Shape of train labels: (60000, 10)
Shape of test labels: (10000, 10)


**网络结构**   
+ 输入层：$28\times 28=784$个神经元
+ 隐含层：$512$个神经元，激活函数relu
+ 输出层：$10$个神经元，激活函数softmax
      
**学习参数**
+ 优化算法：rmsprop
+ 损失函数：交叉熵

In [2]:
from tensorflow.keras import models
from tensorflow.keras import layers

network = models.Sequential()
network.add(layers.Input(shape=(28*28,)))
network.add(layers.Dense(512, activation='relu',kernel_initializer='GlorotUniform'))
network.add(layers.Dense(10, activation='softmax'))

network.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_2 (Dense)             (None, 512)               401920    
                                                                 
 dense_3 (Dense)             (None, 10)                5130      
                                                                 
Total params: 407,050
Trainable params: 407,050
Non-trainable params: 0
_________________________________________________________________


**网络学习和测试**
+ 随机梯度学习参数，每一轮迭代(batch)随机抽取128个训练样本
+ 每个回合(epoch)迭代469轮（60000/128=468.75）
+ 学习5个回合
+ 显示测试样本分类正确率

In [16]:
network.compile(optimizer='rmsprop',loss='categorical_crossentropy',
                metrics=['accuracy'])
network.fit(train_images, train_labels, epochs=5, batch_size=128)
print("\n")

test_loss, test_acc = network.evaluate(test_images, test_labels)
print('Test Accuracy:', test_acc)

Epoch 1/5
469/469 [==============================] - 3s 6ms/step - loss: 0.2635 - accuracy: 0.9237
Epoch 2/5
469/469 [==============================] - 3s 6ms/step - loss: 0.1057 - accuracy: 0.9688
Epoch 3/5
469/469 [==============================] - 3s 6ms/step - loss: 0.0700 - accuracy: 0.9789
Epoch 4/5
469/469 [==============================] - 3s 6ms/step - loss: 0.0515 - accuracy: 0.9846
Epoch 5/5
469/469 [==============================] - 3s 6ms/step - loss: 0.0380 - accuracy: 0.9891


313/313 [==============================] - 0s 1ms/step - loss: 0.0683 - accuracy: 0.9788
Test Accuracy: 0.9787999987602234
